// @shazow/whatsabi/lib.cjs/loaders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultsWithEnv = exports.defaultsWithAPIKeys = exports.defaultSignatureLookup = exports.defaultABILoader = exports.SamczunSignatureLookup = exports.OpenChainSignatureLookup = exports.FourByteSignatureLookup = exports.MultiSignatureLookup = exports.SourcifyABILoader = exports.EtherscanABILoader = exports.MultiABILoader = void 0;
const utils_js_1 = require("./utils.js");
const emptyContractResult = {
    ok: false,
    abi: [],
    name: null,
    evmVersion: "",
    compilerVersion: "",
    runs: 0,
};
class MultiABILoader {
    loaders;
    constructor(loaders) {
        this.loaders = loaders;
    }
    async getContract(address) {
        for (const loader of this.loaders) {
            try {
                const r = await loader.getContract(address);
                if (r && r.abi.length > 0)
                    return Promise.resolve(r);
            }
            catch (error) {
                if (error.status === 404)
                    continue;
                throw error;
            }
        }
        return emptyContractResult;
    }
    async loadABI(address) {
        for (const loader of this.loaders) {
            const r = await loader.loadABI(address);
            if (r.length > 0)
                return Promise.resolve(r);
        }
        return Promise.resolve([]);
    }
}
exports.MultiABILoader = MultiABILoader;
class EtherscanABILoader {
    apiKey;
    baseURL;
    constructor(config) {
        if (config === undefined)
            config = {};
        this.apiKey = config.apiKey;
        this.baseURL = config.baseURL || "https://api.etherscan.io/api";
    }
    async getContract(address) {
        let url = this.baseURL + '?module=contract&action=getsourcecode&address=' + address;
        if (this.apiKey)
            url += "&apikey=" + this.apiKey;
        const r = await (0, utils_js_1.fetchJSON)(url);
        if (r.status === "0") {
            if (r.result === "Contract source code not verified")
                return emptyContractResult;
            throw new Error("Etherscan error: " + r.result);
        }
        const result = r.result[0];
        return {
            abi: JSON.parse(result.ABI),
            name: result.ContractName,
            evmVersion: result.EVMVersion,
            compilerVersion: result.CompilerVersion,
            runs: result.Runs,
            ok: true,
        };
    }
    async loadABI(address) {
        let url = this.baseURL + '?module=contract&action=getabi&address=' + address;
        if (this.apiKey)
            url += "&apikey=" + this.apiKey;
        const r = await (0, utils_js_1.fetchJSON)(url);
        if (r.status === "0") {
            if (r.result === "Contract source code not verified")
                return [];
            throw new Error("Etherscan error: " + r.result);
        }
        return JSON.parse(r.result);
    }
}
exports.EtherscanABILoader = EtherscanABILoader;
function isSourcifyNotFound(error) {
    return (error.message === "Failed to fetch" ||
        error.status === 404);
}
class SourcifyABILoader {
    chainId;
    constructor(config) {
        this.chainId = config?.chainId ?? 1;
    }
    async getContract(address) {
        address = (0, utils_js_1.addressWithChecksum)(address);
        try {
            const r = await (0, utils_js_1.fetchJSON)("https://repo.sourcify.dev/contracts/full_match/" + this.chainId + "/" + address + "/metadata.json");
            return {
                abi: r.output.abi,
                name: r.output.devdoc?.title ?? null,
                evmVersion: r.settings.evmVersion,
                compilerVersion: r.compiler.version,
                runs: r.settings.optimizer.runs,
                ok: true,
            };
        }
        catch (error) {
            if (!isSourcifyNotFound(error))
                throw error;
        }
        try {
            const r = await (0, utils_js_1.fetchJSON)("https://repo.sourcify.dev/contracts/partial_match/" + this.chainId + "/" + address + "/metadata.json");
            return {
                abi: r.output.abi,
                name: r.output.devdoc?.title ?? null,
                evmVersion: r.settings.evmVersion,
                compilerVersion: r.compiler.version,
                runs: r.settings.optimizer.runs,
                ok: true,
            };
        }
        catch (error) {
            if (!isSourcifyNotFound(error))
                throw error;
        }
        return emptyContractResult;
    }
    async loadABI(address) {
        address = (0, utils_js_1.addressWithChecksum)(address);
        try {
            return (await (0, utils_js_1.fetchJSON)("https://repo.sourcify.dev/contracts/full_match/" + this.chainId + "/" + address + "/metadata.json")).output.abi;
        }
        catch (error) {
            if (!isSourcifyNotFound(error))
                throw error;
        }
        try {
            return (await (0, utils_js_1.fetchJSON)("https://repo.sourcify.dev/contracts/partial_match/" + this.chainId + "/" + address + "/metadata.json")).output.abi;
        }
        catch (error) {
            if (!isSourcifyNotFound(error))
                throw error;
        }
        return [];
    }
}
exports.SourcifyABILoader = SourcifyABILoader;
class MultiSignatureLookup {
    lookups;
    constructor(lookups) {
        this.lookups = lookups;
    }
    async loadFunctions(selector) {
        for (const lookup of this.lookups) {
            const r = await lookup.loadFunctions(selector);
            if (r.length > 0)
                return Promise.resolve(r);
        }
        return Promise.resolve([]);
    }
    async loadEvents(hash) {
        for (const lookup of this.lookups) {
            const r = await lookup.loadEvents(hash);
            if (r.length > 0)
                return Promise.resolve(r);
        }
        return Promise.resolve([]);
    }
}
exports.MultiSignatureLookup = MultiSignatureLookup;
class FourByteSignatureLookup {
    async load(url) {
        try {
            const r = await (0, utils_js_1.fetchJSON)(url);
            if (r.results === undefined)
                return [];
            return r.results.map((r) => { return r.text_signature; });
        }
        catch (error) {
            if (error.status === 404)
                return [];
            throw error;
        }
    }
    async loadFunctions(selector) {
        return this.load("https://www.4byte.directory/api/v1/signatures/?hex_signature=" + selector);
    }
    async loadEvents(hash) {
        return this.load("https://www.4byte.directory/api/v1/event-signatures/?hex_signature=" + hash);
    }
}
exports.FourByteSignatureLookup = FourByteSignatureLookup;
class OpenChainSignatureLookup {
    async load(url) {
        try {
            const r = await (0, utils_js_1.fetchJSON)(url);
            if (!r.ok)
                throw new Error("OpenChain API bad response: " + JSON.stringify(r));
            return r;
        }
        catch (error) {
            if (error.status === 404)
                return [];
            throw error;
        }
    }
    async loadFunctions(selector) {
        const r = await this.load("https://api.openchain.xyz/signature-database/v1/lookup?function=" + selector);
        return (r.result.function[selector] || []).map((item) => item.name);
    }
    async loadEvents(hash) {
        const r = await this.load("https://api.openchain.xyz/signature-database/v1/lookup?event=" + hash);
        return (r.result.event[hash] || []).map((item) => item.name);
    }
}
exports.OpenChainSignatureLookup = OpenChainSignatureLookup;
class SamczunSignatureLookup extends OpenChainSignatureLookup {
}
exports.SamczunSignatureLookup = SamczunSignatureLookup;
exports.defaultABILoader = new MultiABILoader([new SourcifyABILoader(), new EtherscanABILoader()]);
exports.defaultSignatureLookup = new MultiSignatureLookup([new OpenChainSignatureLookup(), new FourByteSignatureLookup()]);
function defaultsWithAPIKeys(apiKeys) {
    return defaultsWithEnv(apiKeys);
}
exports.defaultsWithAPIKeys = defaultsWithAPIKeys;
function defaultsWithEnv(env) {
    return {
        abiLoader: new MultiABILoader([
            new SourcifyABILoader({ chainId: env.SOURCIFY_CHAIN_ID && Number(env.SOURCIFY_CHAIN_ID) || undefined }),
            new EtherscanABILoader({ apiKey: env.ETHERSCAN_API_KEY, baseURL: env.ETHERSCAN_BASE_URL }),
        ]),
        signatureLookup: new MultiSignatureLookup([
            new OpenChainSignatureLookup(),
            new FourByteSignatureLookup(),
        ]),
    };
}
exports.defaultsWithEnv = defaultsWithEnv;
